### PR TITLE
Add Ascend NPU native support

### PR DESCRIPTION
**Pull request by FightingZhen**
As proposed in RFC #64 , we refactor device related logics for supporting Ascend NPU and any other third-party platforms in VeOmni.

Detail refactor modifications including:
1. Create new `device.py` file,  define a series of device related apis:
    - `get_device_type`: return device type name, e.g. cuda/npu/cpu
    - `get_device_name`: return device real name, e.g. A100/H100/...
    - `get_torch_device`: return torch attribute based on device type, e.g. torch.cuda/torch.npu
    - `get_device_id`: return current device id
    - `get_nccl_backend`: return NCCL/HCCL backend name
    - `synchronize`: execute synchronize operation
    - `empty_cache`: execute empty cache operation
 2. Refactor `.cuda` related logics with above new defined functions
 3. Refactor "cuda" keywords related logics with above new defined function
 4. Refactor NCCL related logics
 5. Refactor logics related to torch_npu.contrib.transfer_to_npu`

Additionally, to make sure new codes use above functions, we also provide a test script, which will scan all files in `VeOmni/tasks` and `VeOmni/veomni` directory, and find out the new appended `.cuda`, `"cuda"` or `"nccl"` contents. **Hope to append it into VeOmni CI** :)

Based on above modification, we use `tasks/omni/train_qwen2_5_vl.py` for SFT test with following config:

```yaml
model:
  model_path: Qwen/Qwen2.5-VL-7B-Instruct
  attn_implementation: "eager"

data:
  train_path: sharegpt4v/subset
  chat_template: qwen2_5vl
  max_seq_len: 2048
  train_size: 80000000
  source_name: sharegpt4v_pretrain
  enable_multisource: false
  dataloader_type: "native"
  datasets_type: "mapping"

train:
  output_dir: qwen2_5vl_sft
  data_parallel_mode: fsdp1
  wandb_project: qwen2_5vl
  wandb_name: qwen2_5vl
  rmpad: false
  rmpad_with_pos_ids: true
  ulysses_parallel_size: 1
  freeze_vit: false
  lr: 1.0e-5
  lr_decay_style: cosine
  num_train_epochs: 1
  micro_batch_size: 1
  global_batch_size: 32
  max_steps: 2000
  init_device: "npu"
  use_wandb: false
  ckpt_manager: "dcp"
```

Training loss is shown as follows, which looks good :)
ascend_veomni_qwen2_5_vl_sft_loss

### CODE REVIEW

File: veomni/utils/device.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    try:
+        import torch_npu  # noqa: F401
+
+        return torch.npu.is_available()
+    except ImportError:
+        return False
+
+
+is_cuda_available = torch.cuda.is_available()
+is_npu_available = is_torch_npu_available()
```
Comment (Fazziekey) [inline veomni/utils/device.py]: 这个要不还是保留函数is_torch_npu_available()， is_npu_available 的话最好是大写全局变量 IS_NPU_AVAILABLE

Comment (FightingZhen) [reply veomni/utils/device.py]: > 这个要不还是保留函数is_torch_npu_available()， is_npu_available 的话最好是大写全局变量 IS_NPU_AVAILABLE

已解决，还原is_torch_npu_available() veomni原始判断逻辑，同时大写全局变量

File: .github/workflows/pre-commit.yml [hunk 2]
```diff
@@ -32,3 +32,9 @@ jobs:
           pwd && ls -la
           ruff check tasks tests veomni docs
           ruff format --check tasks tests veomni docs
+      - name: Run device api usage normativity in VeOmni/tasks
```
Comment (Fazziekey) [inline .github/workflows/pre-commit.yml]: 这个不要放 pre-commit 里面？单独建立一个 workflow？

File: veomni/optim/optimizer.py [hunk 3]
```diff
@@ -346,8 +346,8 @@ def build_ep_fsdp2_optimizer(
     )
 
     def _build(groups: Sequence[Dict[str, Any]]) -> Optimizer:
-        foreach = False if is_torch_npu_available() else (not fused)
-        fused_ = False if is_torch_npu_available() else fused
+        foreach = False if IS_NPU_AVAILABLE else (not fused)
```
Comment (Fazziekey) [inline veomni/optim/optimizer.py]: 这里还是 is_torch_npu_available()?

File: .github/workflows/pre-commit.yml (hunk 2 above)
Comment (FightingZhen) [reply .github/workflows/pre-commit.yml]: 针对device api校验，已单独建立workflow

File: tasks/omni/train_qwen2_5_vl.py [hunk 4]
```diff
@@ -174,6 +182,7 @@ def main():
         processor=processor,
         chat_template=chat_template,
         position_id_func=position_id_func,
+        source_name=args.data.source_name,
```
Comment (Fazziekey) [inline tasks/omni/train_qwen2_5_vl.py]: 这个 data.source_name 会报错吧，没 入参

File: veomni/ops/group_gemm/utils/benchmark_utils.py [hunk 5]
```diff
@@ -87,7 +91,7 @@ def _report_benchmark_result(
         name = name + " [baseline]"  # ...
 
     msec_per_iter = 1000 / iters_per_sec
-    blog.logging.info(
+    logger.info(
```
Comment (Fazziekey) [inline veomni/ops/group_gemm/utils/benchmark_utils.py]: 这个 blog 别改

File: veomni/utils/arguments.py [hunk 6]
```diff
@@ -140,6 +140,8 @@ class DataArguments:
     train_path: str = field(
         metadata={"help": "Local path/HDFS path of the training data. Use comma to separate multiple datasets."},
     )
+    source_name: str = field(metadata={"help": "Dataset source name, used for preparing corresponding preprocessor."})
+    enable_multisource: bool = field(metadata={"help": "Whether to enable multi-source."})
```
Comment (Fazziekey) [inline veomni/utils/arguments.py]: 这两个不用加

**Comment by onehaitao** [comment]
label ascend
